File: ezkey-cli/src/utils/json-utils.ts
/*
 * Ezkey - Open Source MFA/Passkey Alternative
 *
 * Copyright (c) 2025 Ezkey contributors
 * Licensed under the MIT License. See LICENSE file in the project root for full license information.
 *
 * CLI Component: JSON Utilities
 * Description: JSON parsing, formatting and file handling utilities
 */

import * as fs from 'fs';
import * as path from 'path';

export class JsonUtils {
    /**
     * Pretty print JSON data
     */
    static prettyPrint(data: any, indent: number = 2): string {
        try {
            return JSON.stringify(data, null, indent);
        } catch (error) {
            return 'Error: Unable to serialize data to JSON';
        }
    }

    /**
     * Parse JSON data from string, with error handling
     */
    static parse(jsonString: string): any {
        try {
            return JSON.parse(jsonString);
        } catch (error) {
            throw new Error(`Invalid JSON: ${error instanceof Error ? error.message : String(error)}`);
        }
    }

    /**
     * Load JSON data from file
     * Supports both absolute and relative paths
     */
    static loadFromFile(filePath: string): any {
        try {
            // Resolve relative paths
            const resolvedPath = path.isAbsolute(filePath) 
                ? filePath 
                : path.resolve(process.cwd(), filePath);

            if (!fs.existsSync(resolvedPath)) {
                throw new Error(`File not found: ${resolvedPath}`);
            }

            const content = fs.readFileSync(resolvedPath, 'utf-8');
            return this.parse(content);
        } catch (error) {
            throw new Error(`Failed to load JSON from file ${filePath}: ${error instanceof Error ? error.message : String(error)}`);
        }
    }

    /**
     * Process input value - if it starts with @, treat as file path
     * Otherwise, try to parse as JSON string
     */
    static processInput(input: string): any {
        if (input.startsWith('@')) {
            // File input
            const filePath = input.substring(1);
            return this.loadFromFile(filePath);
        } else {
            // Direct JSON input
            return this.parse(input);
        }
    }

    /**
     * Format output based on pretty print setting
     */
    static formatOutput(data: any, prettyPrint: boolean = true): string {
        if (prettyPrint) {
            return this.prettyPrint(data);
        } else {
            return JSON.stringify(data);
        }
    }

    /**
     * Validate if string is valid JSON
     */
    static isValidJson(str: string): boolean {
        try {
            JSON.parse(str);
            return true;
        } catch {
            return false;
        }
    }

    /**
     * Safely extract value from nested object
     */
    static getValue(obj: any, path: string, defaultValue: any = null): any {
        const keys = path.split('.');
        let current = obj;
        
        for (const key of keys) {
            if (current === null || current === undefined || !(key in current)) {
                return defaultValue;
            }
            current = current[key];
        }
        
        return current;
    }
}

File: ezkey-cli-python/ezkey_cli/utils/json_utils.py
"""
Ezkey - Open Source MFA/Passkey Alternative

Copyright (c) 2025 Ezkey contributors
Licensed under the MIT License. See LICENSE file in the project root for full license information.

CLI Component: JSON Utilities
Description: JSON parsing, formatting and file handling utilities
"""

import json
import os
from pathlib import Path
from typing import Any, Union


class JsonUtils:
    """Utilities for JSON handling and file operations."""
    
    @staticmethod
    def pretty_print(data: Any, indent: int = 2) -> str:
        """Pretty print JSON data."""
        try:
            return json.dumps(data, indent=indent, ensure_ascii=False)
        except (TypeError, ValueError) as e:
            return f"Error: Unable to serialize data to JSON: {str(e)}"
    
    @staticmethod
    def parse(json_string: str) -> Any:
        """Parse JSON data from string, with error handling."""
        try:
            return json.loads(json_string)
        except json.JSONDecodeError as e:
            raise ValueError(f"Invalid JSON: {str(e)}")
    
    @staticmethod
    def load_from_file(file_path: str) -> Any:
        """
        Load JSON data from file.
        Supports both absolute and relative paths.
        """
        try:
            # Resolve relative paths
            path = Path(file_path)
            if not path.is_absolute():
                path = Path.cwd() / file_path
            
            if not path.exists():
                raise FileNotFoundError(f"File not found: {path}")
            
            with open(path, 'r', encoding='utf-8') as f:
                content = f.read()
            
            return JsonUtils.parse(content)
        except Exception as e:
            raise ValueError(f"Failed to load JSON from file {file_path}: {str(e)}")
    
    @staticmethod
    def process_input(input_value: str) -> Any:
        """
        Process input value - if it starts with @, treat as file path.
        Otherwise, try to parse as JSON string.
        """
        if input_value.startswith('@'):
            # File input
            file_path = input_value[1:]
            return JsonUtils.load_from_file(file_path)
        else:
            # Direct JSON input
            return JsonUtils.parse(input_value)
    
    @staticmethod
    def format_output(data: Any, pretty_print: bool = True) -> str:
        """Format output based on pretty print setting."""
        if pretty_print:
            return JsonUtils.pretty_print(data)
        else:
            return json.dumps(data, ensure_ascii=False)
    
    @staticmethod
    def is_valid_json(string: str) -> bool:
        """Validate if string is valid JSON."""
        try:
            json.loads(string)
            return True
        except json.JSONDecodeError:
            return False
    
    @staticmethod
    def get_value(obj: Any, path: str, default: Any = None) -> Any:
        """Safely extract value from nested object using dot notation."""
        keys = path.split('.')
        current = obj
        
        for key in keys:
            if current is None or not isinstance(current, dict) or key not in current:
                return default
            current = current[key]
        
        return current